### ISSUE TITLE
Is PyMuPDF re-entrant / thread-safe?

### ISSUE THREAD

**Issue by MikeTheWatchGuy** [rk700/PyMuPDF#107]
I'm wondering if I might be able to significantly increase my throughput but kicking off a few threads to do my extractions.

While the core my Python code is running in is staying 98% busy,  my i9 CPU is only 4% busy overall.  The remaining cores could be staying quite busy too assuming that the code is thread-safe.

I'll start working on it in the meantime. If I get an answer before you do, I'll post it :-)

**Comment by JorjMcKie** [comment]
Glad you asked that!
We (Ruikai and I) have been faced with the same question a few weeks ago.

Short answer: PyMuPDF is **not** threadable.
More precisely put: it's not in terms of Python. Dropping Python's so-called GIL (global interpreter lock) is in principle possible for MuPDF-based code, but we haven't found a case to justify the (quite considerable) implementation effort against a fairly low performance gain. Please see issue #97 for some considerations on this.

Long answer: there still are options ...
I have played around with the `subprocess` module.  Look at this script:

```
import time
import os
import subprocess

print(__file__)
pdfdir = "where all the PDFs live ..."
pdfs = os.listdir(pdfdir)

t0 = time.clock()
cnt = 0
threads = []
for i, pdf in enumerate(pdfs):
    if not pdf.endswith(".pdf"):
        continue
    cnt += 1
    pth = os.path.join(pdfdir, pdf)
    t = subprocess.Popen("python extract_img2.py %s" % pth,  shell = True)
    threads.append(t)    

# now wait for all the guys to finish ...
for t in threads:
    t.wait()

t1 = time.clock()
print("time needed for %i documents: %g" % (cnt, t1-t0))
print("average time %g" % ((t1-t0)/cnt))
```

This script should really exploit all CPU cores you have.
I am doing similar stuff when I re-generate PyMuPDF: because I need to create binaries for all 5 Python versions separately, and one for each bitness, I trigger 10 independent compiles with this method.
It saves me about 60% overall elapsed time on my 8 core 4.0 GHz machine: it is now 40 seconds, where it used to be up to 2 minutes before.

**Comment by JorjMcKie** [comment]
PS:
Of course `extract_img2.py` output would have to make sure that image file names are unique across PDFs ...

**Comment by MikeTheWatchGuy** [comment]
I'm working on a program now that is multi-threaded.  I'm protecting my critical regions using the Threading module's `threading.lock() `object.  It's been working great for me.

Perhaps I should look at multi-processed if multi-threading is going to be super-tricky.

If PyMyPDF isn't threadsafe on it's own, then I could put one of these threading lock's around the calls to the library.  That should do the trick unless there's some other kind of problem other than a 'race condition' to worry about.

**Comment by JorjMcKie** [comment]
Oh, you can use PyMuPDF in a multi-threading environment without problem - it is **safe** in that sense.
However, MuPDF is not reentrant and hence mutiple uses (e.g. via `threading.lock()`) of it get queued behind each other automatically, thus annihilating any performance gain.
What we have been considering on occasion of #97, was triggering multiple copies of our underlying MuPDF library. There is one critical memory area MuPDF is using for all of its work, the "global context". This area can not be shared, so each task must be given its owns global context. This would mean that ominous "considerable effort" I had been referring to about earlier: today there is exactly one global context, which we therefore could elegantly keep out of sight of PyMuPDF's users.
As it stands, the only option now is starting multiple processes, each with its own Python interpreter.

**Comment by MikeTheWatchGuy** [comment]
Let me make sure I have this correct....

It's OK for to have multiple Documents open at a time using fitz:
`doc = fitz.open('pdf', b)`

If so, then I will wrap each of my calls with locks such as this:
```
threading.acquire()
doc._getXrefLength()
threading.release()
```
and this:
```
threading.acquire()
text = doc._getObjectString(i)
threading.release()

```

I can assure you that gains will not be 'annihilated' if my threads block on these calls.  The overhead of the switching won't be that bad compared to the current situation of a single thread with 19 idle cores that could be running.  OK, I think I can assure you :-)

Please let me know if multiple docs won't be a problem.

**Comment by JorjMcKie** [comment]
It is no problem to have open several files in PyMuPDF in parallel (assured! ;-)). This is only limited by memory.
And your above code snippets won't do anything bad either (crashes or so).
I just believe there will be no throughput gain because of my argument above.

19 cores ... WOW .. you have a hell of a machine. And I thought my 8 core 32 GB 4.0 GHz isn't really bad.

**Comment by JorjMcKie** [comment]
Can we close this?
A labelled the equivalent issue #97 as **enhancement** that be **won't fix*. I am afraid, I do not have a different answer today.

**Comment by JorjMcKie** [comment]
@rk700 - did you come to a different position on this in the meantime?

**Comment by MikeTheWatchGuy** [comment]
I fully coded up a multi-threaded version of the Extract test code I integrated into my code a couple weeks back.  I added locks around all of my Fitz calls and turned loose from 1 to a couple hundred threads.  I got almost identical results as single threaded.  :-(  Lesson learned.

Following Jorj's excellent lead, I then bumped UP a level to multi-processing instead of multi-threading.  The results have been BLISTERING.  It takes roughly 1.5 seconds to extract all of the images from my test PDFs.  I'm running on a core i9-7900X @ 3.3 GHz with 64 GB RAM.  The files were stored on an M.2 drive to give an added boost to performance.

I just finished extracting 15,933 images from 2,409 PDF files that were all in the range of 1 to 2 MB.

Running 18 sub-processes I was able to process 1,841 PDF files that were all > 2 MB that resulted in 24,212 images took 15:36 (that's minutes not hours 👍 That's 1.36 PDFs / second.

The way I did this was using the` subprocess.Popen ` call as demonstrated in Jorj's code.  I started 18 Threads which in turn kicked off these subprocesses which processed a single PDF file.  

There may be even more to gain by sending a list of files rather than single files.  It's hard to argue too much with less than 1.5 seconds per PDF.

**Comment by JorjMcKie** [comment]
Again my congratulations!
You brought up a very valid consideration, namely should you pass one or multiple files to each kicked-off process?

# Some thoughts on this
There is certainly an overhead incurred with each of

1. starting a Windows process
2. starting Python
3. importing fitz

It could be worthwhile minimizing the impact of these overheads on the overall elapsed time (which is the ultimate goal).

Here are some test results on my machine (4.0 GHz, Win10 PRO).

## Python 3.6 64bit
Importing fitz (point 3. above) takes about 0.0054 s
I cannot separate points 1. and 2. above, but the combined time of **1. to 3.** have been 0.094 s.
Therefore, 1 + 2 should be 0.094 - 0.0054, or roughly 0.089 s.
All this seems very low.

Taking the results of @MikeTheWatchGuy:
Those 1.841 files have been processed by 18 cores, i.e. 102 files by each one. The overall elapsed time of 15:36 minutes = 936 seconds should be assumed to be the time **that each core was busy**. So, the average elapsed **core time** of each file was 936 / 102 = 9.2 seconds, and the accumulated overhead per core was 0.094 * 102, i.e. around 10 seconds. 

So overall:
Getting rid of **all overhead** could theoreticall reduce the overall elapsed from 15:36 to 15:26.
A more realistic savings number is probably 2 to 5 seconds, because you probably break the 1.841 into subsets of 10 or 20 files (you cannot predict the perfect equal load per core).

My recommendation: the effort to add this level of complexity is not worth the to-be-expected benefit.

Any big flaws in this argument?

**Comment by JorjMcKie** [comment]
I was just thinking:
The above is still missing one overhead component, namely the post-process clean-up time. This of cause needs to happen as well, before a CPU core can be used again. I currently have no data to tell what that might be.
It should be fair to assume a similar time as for the process start-up time. This would increase the theoretical time saving to by another 10 seconds - still not overwhelming.

**Comment by MikeTheWatchGuy** [comment]
Your analysis is spot on.   You've got excellent instincts FOLLOWED by immediate action.  Man, do you ever pause?  

My original design of the multi-threaded was to divide up the files into equal sections.  But then I realized I could just grab them off a list one by one.  This is still what I'm doing in my sub-processed based one does. Each thread gets a file to process, kicks off the subprocess, waits for it to end, lather, rinse, repeat.

Thanks for the hard numbers on the overhead and potential savings.  Shockingly small.  I must still be thinking in terms of 12 MHz clock speeds systems.  That exponential thing sure creeps up on you.

**Comment by MikeTheWatchGuy** [comment]
I was wondering what your thoughts are on 'import creep' (dunno if it's a term, but it is now).

I'm talking specifically about the additions I make to my simple utilities.  I have not yet submitted anything to GitHub so it's not a formal problem, yet.

There are two examples I wish to use. One is a simple GUI that I created for my utilities that displays a Confirmation or Message box.  I encapsulated in my own Module I named MikesSimpleGUI.

I also have been adding tqdm to almost all of my recent utilities.  This is because I'm processing large numbers of files plus it's a fantastic way to get basic statistics.  For example, my latest version of Jorj's de-duplication utility showed these stats:
Exited GUI... Beginning de-dupe
100%|██████████| 12030/12030 [00:14<00:00, 817.80it/s] 
Removed 0 duplicate or small files from a total of 12029.

It's handy to know that I can rip thought 817 files per second and compute an SHA hash for each, etc.

Is there a way to use something similar to #define to turn on and off these added functionality that is perhaps unique to my setup?


I tried to attach my DeDupe.py file, but was unable... so, I'm going to paste it here (I'm sorry if this is a complete no-no in Git)
```

import hashlib
import os
import MikesSimpleGUI as GUI
from tkinter import *
from tkinter import filedialog
from tqdm import tqdm

INPUT_PATH = 'A:\\'


# ====____====____==== CLASS DeDupeGUI ====____====____==== #
# Contains all of the variable and functions to implement   #
# the DeDupe user interface                                 #
# --------------------------------------------------------- #
class DeDupeGUI:
    def __init__(self, master, default_source_path):
        self._master = master
        self._source_stringvar = ''

        self.submit_pressed = False
        self.source_folder = ''

        # ===================================== CREATE the widgets==========================================#

        # =========== Set TITLE for main window ==========#
        master.title('PDF Image Extract')

        # =========== Create Labels explaining what app does ==========#
        Label(master, text="PDF Image Extraction").grid(row=1, sticky=W)
        Label(master, text="Choose source folder").grid(row=2, sticky=W)

        # =========== Create Label Widgets for Text Entry ==========#
        self.source_label = Label(master, text="Source Folder")

        # =========== GUI values stored in 'stringvar'==========#
        self._source_stringvar = StringVar()
        self._source_stringvar.set(default_source_path)
        self.source_folder = default_source_path

        # =========== Create TEXT Entry Widgets ==========#
        self.text_entry_source = Entry(master, width=80, textvariable=self._source_stringvar)

        # =========== Create Button Widgets ==========#
        self.browse_source_button = Button(master, text="Browse", command=self.change_source_button)

        self.submit_button = Button(master, text="Submit", command=self.submit_button)

        # ===================================== LAYOUT the widgets==========================================#
        self.source_label.grid(row=3, column=0, sticky=W)  ### Label Widget    ###
        self.text_entry_source.grid(row=4, column=0, sticky=W)  ### Text ENTRY Widget    ###
        self.browse_source_button.grid(row=4, column=2, sticky=E)  ### Browse Button Widget    ###

        self.submit_button.grid(row=7)  ### Submit Button Widget    ###

    # ===================================== CALLBACKS for widgets=======================================#

    # ------------BUTTON Source -----------===========-----------#
    def change_source_button(self):
        filename = filedialog.askdirectory()

        if filename is '':
            print('User Cancelled')
        else:
            self._source_stringvar.set(filename)
            self.source_folder = filename

        print('The new source folder is ' + self.source_folder)
        # self.text_entry_source.insert(END, filename)
        return


    # ------------BUTTON Submit -----------===========-----------#
    def submit_button(self):
        self.source_folder = self._source_stringvar.get()  # grab values from text boxes before exiting

        print('SUBMITTING....\nSource= {}\n'.format(self.source_folder))
        self.submit_pressed = True  # User pressed Submit button to exit (versus closing window or cancelling)

        self._master.quit()  # exit the GUI entirely
        return


# ====____====____==== FUNCTION DeDuplicate_folder(path) ====____====____==== #
# Function to de-duplicate the folder passed in                               #
# --------------------------------------------------------------------------- #
def DeDuplicate_folder(path):
    pngdir = path
    pngfiles = os.listdir(pngdir)
    shatab = []
    dups = 0
    total = 0
    small = 2048
    pbar = None

    pbar = tqdm(total=len(pngfiles), file=sys.stdout)
    for f in pngfiles:
        pbar.update(1)

        if not f.endswith(".png") and not f.endswith(".jpg"):
            continue
        total += 1
        fname = os.path.join(pngdir, f)
        x = open(fname, "rb").read()

        if len(x) <= small:
            dups += 1
            # print('Removing due to too small. Size = {}  ===> {}'.format(len(x), fname))
            os.remove(fname)
            continue

        m = hashlib.sha256()
        m.update(x)
        f_sha = m.digest()
        if f_sha in shatab:
            os.remove(fname)
            # print('Removing due to duplicate ===> {}'.format(fname))
            dups += 1
            continue
        shatab.append(f_sha)
    pbar.close()

    print("Removed %i duplicate or small files from a total of %i." % (dups, total))

# ====____====____==== Pseudo-MAIN program ====____====____==== #
# Thie is our main-alike piece of code                          #
#   + Starts up the GUI                                         #
#   + Gets values from GUI                                      #
#   + Runs DeDupe_folder based on GUI inputs                    #
# ------------------------------------------------------------- #
if __name__ == '__main__':
    source = INPUT_PATH
    root = Tk()
    my_gui = DeDupeGUI(root, source)
    root.mainloop()  # Start up the GUI!
    print('Exited GUI... Beginning de-dupe')
    if (my_gui.submit_pressed == True):  # SUBMIT pressed
        root.destroy()  # Back from the GUI so close the window
        DeDuplicate_folder(my_gui.source_folder)
    else:
        GUI.MessageBox('Cancelling', '*** Cancelling ***')
        print('Exiting without doing de-duplicate')
    exit(0)

```
